File: bindings/cpp/tests/emit_wav.cpp
// Minimal C++ emitter to match golden_output.wav configuration
#include "../ucra_cpp.hpp"
#include <fstream>
#include <vector>
#include <cstdint>

static void write_wav_file(const char* filename, const std::vector<float>& pcm,
                           uint32_t sample_rate, uint32_t channels) {
    std::ofstream file(filename, std::ios::binary);
    uint32_t data_size = static_cast<uint32_t>(pcm.size() * sizeof(float));
    uint32_t file_size = 36 + data_size;
    uint16_t bits_per_sample = 32;
    uint32_t byte_rate = sample_rate * channels * (bits_per_sample / 8);
    uint16_t block_align = channels * (bits_per_sample / 8);

    file.write("RIFF", 4);
    file.write(reinterpret_cast<const char*>(&file_size), 4);
    file.write("WAVE", 4);
    file.write("fmt ", 4);
    uint32_t fmt_size = 16;
    uint16_t audio_format = 3; // IEEE float
    file.write(reinterpret_cast<const char*>(&fmt_size), 4);
    file.write(reinterpret_cast<const char*>(&audio_format), 2);
    file.write(reinterpret_cast<const char*>(&channels), 2);
    file.write(reinterpret_cast<const char*>(&sample_rate), 4);
    file.write(reinterpret_cast<const char*>(&byte_rate), 4);
    file.write(reinterpret_cast<const char*>(&block_align), 2);
    file.write(reinterpret_cast<const char*>(&bits_per_sample), 2);
    file.write("data", 4);
    file.write(reinterpret_cast<const char*>(&data_size), 4);
    file.write(reinterpret_cast<const char*>(pcm.data()), data_size);
}

int main() {
    try {
        ucra::Engine engine; // default options
        ucra::RenderConfig cfg(44100, 1, 512);
        ucra::NoteSegment note(0.0, 2.0, 67, 120, "sol");
        cfg.add_note(std::move(note));
        auto res = engine.render(cfg);
        write_wav_file("cpp_sample_output.wav", res.pcm(), res.sample_rate(), res.channels());
        return 0;
    } catch (...) {
        return 1;
    }
}

File: bindings/rust/ucra/examples/emit_wav.rs
use std::fs::File;
use std::io::Write;

fn write_wav_float32(path: &str, pcm: &[f32], sample_rate: u32, channels: u32) -> std::io::Result<()> {
    let mut f = File::create(path)?;
    let data_size = (pcm.len() * 4) as u32;
    let file_size = data_size + 36;
    f.write_all(b"RIFF")?;
    f.write_all(&file_size.to_le_bytes())?;
    f.write_all(b"WAVE")?;
    f.write_all(b"fmt ")?;
    f.write_all(&16u32.to_le_bytes())?;
    f.write_all(&3u16.to_le_bytes())?; // IEEE float
    f.write_all(&(channels as u16).to_le_bytes())?;
    f.write_all(&sample_rate.to_le_bytes())?;
    let byte_rate = sample_rate * channels * 4;
    f.write_all(&byte_rate.to_le_bytes())?;
    let block_align: u16 = (channels * 4) as u16;
    f.write_all(&block_align.to_le_bytes())?;
    f.write_all(&32u16.to_le_bytes())?;
    f.write_all(b"data")?;
    f.write_all(&data_size.to_le_bytes())?;
    let bytes: &[u8] = unsafe { std::slice::from_raw_parts(pcm.as_ptr() as *const u8, pcm.len() * 4) };
    f.write_all(bytes)?;
    Ok(())
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let mut eng = ucra::Engine::new()?;
    let note = ucra::NoteSegment { start_sec: 0.0, duration_sec: 2.0, midi_note: 67, velocity: 120, lyric: None, f0_override: None, env_override: None };
    let notes = [note];
    let cfg = ucra::RenderConfig::new(44100, 1, &notes);
    let out = eng.render(&cfg)?;
    let pcm = out.pcm().ok_or("no pcm")?;
    write_wav_float32("rust_sample_output.wav", pcm, out.sample_rate(), out.channels())?;
    println!("Wrote rust_sample_output.wav");
    Ok(())
}
